socorro/unittest/lib/test_libjsonschema.py

```diff
@@ -0,0 +1,438 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import jsonschema
+import pytest
+
+from socorro.lib.libjsonschema import (
+    InvalidDocumentError,
+    InvalidSchemaError,
+    JsonSchemaReducer,
+    resolve_references,
+)
+
+
+@pytest.mark.parametrize(
+    "schema, expected",
+    [
+        ({}, {}),
+        # Schemas with no references are returned as is
+        (
+            {"properties": {"color": {"type": "string"}}, "type": "object"},
+            {"properties": {"color": {"type": "string"}}, "type": "object"},
+        ),
+        (
+            {
+                "properties": {
+                    "color": {
+                        "properties": {
+                            "name": {"type": "string"},
+                            "hex": {"type": "string"},
+                        },
+                        "type": "object",
+                    },
+                },
+                "type": "object",
+            },
+            {
+                "properties": {
+                    "color": {
+                        "properties": {
+                            "name": {"type": "string"},
+                            "hex": {"type": "string"},
+                        },
+                        "type": "object",
+                    },
+                },
+                "type": "object",
+            },
+        ),
+        (
+            {"items": {"type": "string"}, "type": "array"},
+            {"items": {"type": "string"}, "type": "array"},
+        ),
+        # Expand references in objects
+        (
+            {
+                "definitions": {
+                    "color_name": {
+                        "type": "string",
+                        "description": "color name",
+                    },
+                    "color": {
+                        # Note: This gets stomped on
+                        "description": "ignored",
+                        "properties": {
+                            "name": {
+                                "$ref": "#/definitions/color_name",
+                                "permissions": ["public"],
+                            },
+                            "cost": {
+                                "permissions": ["protected"],
+                                "type": "integer",
+                            },
+                        },
+                        "permissions": ["ignored"],
+                        "type": "object",
+                    },
+                },
+                "properties": {
+                    "color": {
+                        "description": "the color used here",
+                        "$ref": "#/definitions/color",
+                        "permissions": ["protected"],
```

Comment:
jsonschema doesn't have "permissions" keys and we don't use it that way, so we should nix the permissions bits here.